frontend/src/utilities/input.ts

```diff
@@ -48,24 +52,76 @@ export async function handleKeyDown(e: KeyboardEvent) {
 	}
 }
 
-export async function handleKeyUp(e: KeyboardEvent) {
+export async function onKeyUp(e: KeyboardEvent) {
 	if (shouldRedirectKeyboardEventToBackend(e)) {
 		e.preventDefault();
 		const modifiers = makeModifiersBitfield(e.ctrlKey, e.shiftKey, e.altKey);
 		(await wasm).on_key_up(e.key, modifiers);
 	}
 }
 
-export async function handleMouseDown(e: MouseEvent) {
+// Mouse events
+
+export async function onMouseMove(e: MouseEvent) {
+	if (!e.buttons) viewportMouseInteractionOngoing = false;
+
+	const modifiers = makeModifiersBitfield(e.ctrlKey, e.shiftKey, e.altKey);
+	(await wasm).on_mouse_move(e.clientX, e.clientY, e.buttons, modifiers);
+}
+
+export async function onMouseDown(e: MouseEvent) {
 	const target = e.target && (e.target as HTMLElement);
-	const clickedInsideDialog = target && target.closest(".dialog-modal .floating-menu-content");
+	const inCanvas = target && target.closest(".canvas");
+	const inDialog = target && target.closest(".dialog-modal .floating-menu-content");
 
-	if (dialogIsVisible() && !clickedInsideDialog) {
-		dismissDialog();
+	// Block middle mouse button auto-scroll mode
+	if (e.button === 1) e.preventDefault();
 
+	if (dialogIsVisible() && !inDialog) {
+		dismissDialog();
 		e.preventDefault();
 		e.stopPropagation();
 	}
+
+	if (inCanvas) viewportMouseInteractionOngoing = true;
+
+	if (viewportMouseInteractionOngoing) {
+		const modifiers = makeModifiersBitfield(e.ctrlKey, e.shiftKey, e.altKey);
```

Comment:
why don't you just pass the event? I guess for cleanliness but probably performs worse